File: internal/types/feed.go
package types

import (
	"fmt"
	"time"

	"github.com/aaronwald/ssmd/internal/utils"
)

// FeedType represents the type of data feed
type FeedType string

const (
	FeedTypeWebSocket FeedType = "websocket"
	FeedTypeREST      FeedType = "rest"
	FeedTypeMulticast FeedType = "multicast"
)

// FeedStatus represents the operational status of a feed
type FeedStatus string

const (
	FeedStatusActive     FeedStatus = "active"
	FeedStatusDeprecated FeedStatus = "deprecated"
	FeedStatusDisabled   FeedStatus = "disabled"
)

// AuthMethod represents authentication methods
type AuthMethod string

const (
	AuthMethodAPIKey AuthMethod = "api_key"
	AuthMethodOAuth  AuthMethod = "oauth"
	AuthMethodMTLS   AuthMethod = "mtls"
	AuthMethodNone   AuthMethod = "none"
)

// TransportProtocol represents the network transport
type TransportProtocol string

const (
	TransportWSS       TransportProtocol = "wss"
	TransportHTTPS     TransportProtocol = "https"
	TransportMulticast TransportProtocol = "multicast"
	TransportTCP       TransportProtocol = "tcp"
)

// MessageProtocol represents the wire format
type MessageProtocol string

const (
	MessageJSON     MessageProtocol = "json"
	MessageITCH     MessageProtocol = "itch"
	MessageFIX      MessageProtocol = "fix"
	MessageSBE      MessageProtocol = "sbe"
	MessageProtobuf MessageProtocol = "protobuf"
)

// Protocol represents connection and message protocols
type Protocol struct {
	Transport TransportProtocol `yaml:"transport"`
	Message   MessageProtocol   `yaml:"message"`
	Version   string            `yaml:"version,omitempty"`
}

// SiteType represents the type of capture location
type SiteType string

const (
	SiteTypeCloud  SiteType = "cloud"
	SiteTypeColo   SiteType = "colo"
	SiteTypeOnPrem SiteType = "on_prem"
)

// CaptureLocation represents where feed data is captured
type CaptureLocation struct {
	Site     string   `yaml:"site"`
	Type     SiteType `yaml:"type"`
	Provider string   `yaml:"provider,omitempty"`
	Region   string   `yaml:"region,omitempty"`
	Clock    string   `yaml:"clock,omitempty"` // future: ptp, gps, ntp, local
}

// Feed represents a market data feed configuration
type Feed struct {
	Name             string            `yaml:"name"`
	DisplayName      string            `yaml:"display_name,omitempty"`
	Type             FeedType          `yaml:"type"`
	Status           FeedStatus        `yaml:"status,omitempty"`
	CaptureLocations []CaptureLocation `yaml:"capture_locations,omitempty"`
	Versions         []FeedVersion     `yaml:"versions"`
	Calendar         *Calendar         `yaml:"calendar,omitempty"`
}

// GetName returns the feed name (implements utils.Named)
func (f *Feed) GetName() string { return f.Name }

// FeedVersion represents a version of feed configuration
type FeedVersion struct {
	Version                 string            `yaml:"version"`
	EffectiveFrom           string            `yaml:"effective_from"`
	EffectiveTo             string            `yaml:"effective_to,omitempty"`
	Protocol                Protocol          `yaml:"protocol"`
	Endpoint                string            `yaml:"endpoint"`
	AuthMethod              AuthMethod        `yaml:"auth_method,omitempty"`
	RateLimitPerSecond      int               `yaml:"rate_limit_per_second,omitempty"`
	MaxSymbolsPerConnection int               `yaml:"max_symbols_per_connection,omitempty"`
	SupportsOrderbook       bool              `yaml:"supports_orderbook,omitempty"`
	SupportsTrades          bool              `yaml:"supports_trades,omitempty"`
	SupportsHistorical      bool              `yaml:"supports_historical,omitempty"`
	ParserConfig            map[string]string `yaml:"parser_config,omitempty"`
}

// GetEffectiveFrom implements the Versioned interface
func (v FeedVersion) GetEffectiveFrom() string {
	return v.EffectiveFrom
}

// Calendar represents trading hours and holidays
type Calendar struct {
	Timezone        string `yaml:"timezone,omitempty"`
	HolidayCalendar string `yaml:"holiday_calendar,omitempty"`
	OpenTime        string `yaml:"open_time,omitempty"`
	CloseTime       string `yaml:"close_time,omitempty"`
}

// Validate checks if the feed configuration is valid
func (f *Feed) Validate() error {
	if f.Name == "" {
		return fmt.Errorf("feed name is required")
	}

	// Validate feed type
	switch f.Type {
	case FeedTypeWebSocket, FeedTypeREST, FeedTypeMulticast:
		// valid
	default:
		return fmt.Errorf("invalid feed type: %s (must be websocket, rest, or multicast)", f.Type)
	}

	// Validate status if set
	if f.Status != "" {
		switch f.Status {
		case FeedStatusActive, FeedStatusDeprecated, FeedStatusDisabled:
			// valid
		default:
			return fmt.Errorf("invalid feed status: %s (must be active, deprecated, or disabled)", f.Status)
		}
	}

	// Must have at least one version
	if len(f.Versions) == 0 {
		return fmt.Errorf("feed must have at least one version")
	}

	// Validate versions
	seenDates := make(map[string]bool)
	for i, v := range f.Versions {
		if v.Version == "" {
			return fmt.Errorf("version %d: version identifier is required", i)
		}
		if v.EffectiveFrom == "" {
			return fmt.Errorf("version %s: effective_from is required", v.Version)
		}
		// Validate date format
		if _, err := time.Parse("2006-01-02", v.EffectiveFrom); err != nil {
			return fmt.Errorf("version %s: invalid effective_from date format (expected YYYY-MM-DD): %w", v.Version, err)
		}
		// Check for overlapping dates
		if seenDates[v.EffectiveFrom] {
			return fmt.Errorf("version %s: duplicate effective_from date %s", v.Version, v.EffectiveFrom)
		}
		seenDates[v.EffectiveFrom] = true

		if v.Endpoint == "" {
			return fmt.Errorf("version %s: endpoint is required", v.Version)
		}
	}

	return nil
}

// GetVersionForDate returns the active version for a given date
func (f *Feed) GetVersionForDate(date time.Time) *FeedVersion {
	dateStr := date.Format("2006-01-02")
	sorted := SortVersionsDesc(f.Versions)

	// Find the first version where effective_from <= date and (effective_to is empty or >= date)
	for i := range sorted {
		if sorted[i].EffectiveFrom <= dateStr {
			if sorted[i].EffectiveTo == "" || sorted[i].EffectiveTo >= dateStr {
				return &sorted[i]
			}
		}
	}

	return nil
}

// GetLatestVersion returns the most recent version
func (f *Feed) GetLatestVersion() *FeedVersion {
	if len(f.Versions) == 0 {
		return nil
	}

	sorted := SortVersionsDesc(f.Versions)
	return &sorted[0]
}

// LoadFeed loads a feed from a YAML file
func LoadFeed(path string) (*Feed, error) {
	feed, err := utils.LoadYAML[Feed](path)
	if err != nil {
		return nil, fmt.Errorf("failed to load feed: %w", err)
	}

	// Set default status
	if feed.Status == "" {
		feed.Status = FeedStatusActive
	}

	return feed, nil
}

// SaveFeed saves a feed to a YAML file
func SaveFeed(feed *Feed, path string) error {
	return utils.SaveYAML(feed, path)
}

// LoadAllFeeds loads all feeds from a directory
func LoadAllFeeds(dir string) ([]*Feed, error) {
	return utils.LoadAllYAML(dir, LoadFeed)
}

File: ssmd-agent/src/lib/types/feed.ts
// Feed types and Zod schemas - ported from internal/types/feed.go
import { z } from "zod";

// Enums
export const FeedTypeEnum = z.enum(["websocket", "rest", "multicast"]);
export const FeedStatusEnum = z.enum(["active", "deprecated", "disabled"]);
export const AuthMethodEnum = z.enum(["api_key", "oauth", "mtls", "none"]);
export const TransportProtocolEnum = z.enum(["wss", "https", "multicast", "tcp"]);
export const MessageProtocolEnum = z.enum(["json", "itch", "fix", "sbe", "protobuf"]);
export const SiteTypeEnum = z.enum(["cloud", "colo", "on_prem"]);

// Protocol schema
export const ProtocolSchema = z.object({
  transport: TransportProtocolEnum,
  message: MessageProtocolEnum,
  version: z.string().optional(),
});

// Capture location schema
export const CaptureLocationSchema = z.object({
  site: z.string(),
  type: SiteTypeEnum,
  provider: z.string().optional(),
  region: z.string().optional(),
  clock: z.string().optional(),
});

// Calendar schema
export const CalendarSchema = z.object({
  timezone: z.string().optional(),
  holiday_calendar: z.string().optional(),
  open_time: z.string().optional(),
  close_time: z.string().optional(),
});

// Feed version schema
export const FeedVersionSchema = z.object({
  version: z.string(),
  effective_from: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Date must be YYYY-MM-DD"),
  effective_to: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
  protocol: ProtocolSchema,
  endpoint: z.string().url(),
  auth_method: AuthMethodEnum.optional(),
  rate_limit_per_second: z.number().int().positive().optional(),
  max_symbols_per_connection: z.number().int().positive().optional(),
  supports_orderbook: z.boolean().optional(),
  supports_trades: z.boolean().optional(),
  supports_historical: z.boolean().optional(),
  parser_config: z.record(z.string()).optional(),
});

// Main feed schema
export const FeedSchema = z.object({
  name: z.string().min(1, "Feed name is required"),
  display_name: z.string().optional(),
  type: FeedTypeEnum,
  status: FeedStatusEnum.default("active"),
  capture_locations: z.array(CaptureLocationSchema).optional(),
  versions: z.array(FeedVersionSchema).min(1, "Feed must have at least one version"),
  calendar: CalendarSchema.optional(),
});

// Type exports
export type FeedType = z.infer<typeof FeedTypeEnum>;
export type FeedStatus = z.infer<typeof FeedStatusEnum>;
export type AuthMethod = z.infer<typeof AuthMethodEnum>;
export type TransportProtocol = z.infer<typeof TransportProtocolEnum>;
export type MessageProtocol = z.infer<typeof MessageProtocolEnum>;
export type SiteType = z.infer<typeof SiteTypeEnum>;
export type Protocol = z.infer<typeof ProtocolSchema>;
export type CaptureLocation = z.infer<typeof CaptureLocationSchema>;
export type Calendar = z.infer<typeof CalendarSchema>;
export type FeedVersion = z.infer<typeof FeedVersionSchema>;
export type Feed = z.infer<typeof FeedSchema>;

// Helper functions

/**
 * Get the active version for a given date
 */
export function getVersionForDate(feed: Feed, date: Date): FeedVersion | null {
  const dateStr = date.toISOString().split("T")[0];
  const sorted = [...feed.versions].sort((a, b) =>
    b.effective_from.localeCompare(a.effective_from)
  );

  for (const v of sorted) {
    if (v.effective_from <= dateStr) {
      if (!v.effective_to || v.effective_to >= dateStr) {
        return v;
      }
    }
  }

  return null;
}

/**
 * Get the latest (most recent) version
 */
export function getLatestVersion(feed: Feed): FeedVersion | null {
  if (feed.versions.length === 0) return null;

  const sorted = [...feed.versions].sort((a, b) =>
    b.effective_from.localeCompare(a.effective_from)
  );

  return sorted[0];
}
